using_restricted_array.rb

```diff
@@ -7,41 +7,94 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  indexes = 0
+  until array[indexes] == nil
+    indexes += 1
+  end
+  indexes
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  new_array = []
+  length = length(array)
+  index = 0
+  until index = length - 1
+    new_array << array[index]
+    puts array[index]
+    index += 1
+
+  end
+
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  (0 .. (length - 1)).each do |index|
+    if array[length] == value_to_find
+      return true
+      exit
+    end
+  end
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  largest = 0
+  (0 .. (length - 1)).each do |index|
+    if array[index] > largest
+      largest = array[index]
+    end
+  end
+  largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  smallest = 10000
+  (0 .. (length - 1)).each do |index|
+    if array[index] < smallest
+      smallest = array[index]
+    end
+  end
+  smallest
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  index = length - 1
+  new_array = []
+  until index < 0
+    new_array << array[index]
+    index -= 1
+  end
+  new_array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  lowest_index = 0
+  highest_index = (length - 1)
+  middle_index = (lowest_index + highest_index)/2
+  until lowest_index >= highest_index
+    if value_to_find == array[lowest_index] || array[highest_index] || middle_index
+      true
+      exit
+    elsif array[lowest_index] < value_to_find && array[highest_index] > value_to_find && array[middle_index] < value_to_find
+      highest_index -=1
```

Comment:
You'll also need to compute middle_index for each of the elsif blocks since you are computing middle_index outside the loop before this.